### PR TITLE
chore(middleware-flexible-checksums): use RequestChecksumCalculation and ResponseChecksumValidation

### PR DESCRIPTION
**Pull request by trivikr**
### Issue
* Internal JS-5396
* Alternative to https://github.com/aws/aws-sdk-js-v3/pull/6484 with `flexibleChecksumsInterceptorMiddleware`

### Description
* Uses values in RequestChecksumCalculation and ResponseChecksumValidation to decide flexible checksums algorithm.
* Adds `flexibleChecksumsInterceptorMiddleware` to populate `input[requestValidationModeMember]`
* Deprecates MD5, and sets CRC32 as default.

### Testing
Unit testing in CI

Also, verified that checksums are computed by default
```js
import { S3 } from "../aws-sdk-js-v3/clients/client-s3/dist-cjs/index.js";
import { NodeHttpHandler } from "@smithy/node-http-handler";
import { equal } from "assert";

// Prints checksum headers for request and response.
class CustomHandler extends NodeHttpHandler {
  constructor() {
    super();
  }

  printChecksumHeaders(prefix, headers) {
    for (const [header, value] of Object.entries(headers)) {
      if (header.startsWith("x-amz-checksum-") || header.startsWith("x-amz-sdk-checksum-")) {
        console.log(`${prefix}['${header}']: '${value}'`);
      }
    }
  }

  async handle(request, options) {
    this.printChecksumHeaders("request", request.headers);
    const response = await super.handle(request, options);
    this.printChecksumHeaders("response", response.response.headers);
    return response;
  }
}

// WHEN_SUPPORTED is default.
const client = new S3({
  requestHandler: new CustomHandler(),
});
const Bucket = "test-flexible-checksums-v2"; // Replace with your test bucket name.
const Key = "hello-world.txt";
const Body = "Hello World"; // Replace with the content you want to test.

console.log("Put Object");
// ChecksumAlgorithm is not explicitly required to be specified.
await client.putObject({ Bucket, Key, Body });

console.log("\nGet Object");
// ChecksumMode is not explicitly required to be specified.
const response = await client.getObject({ Bucket, Key });

equal(Body, await response.Body.transformToString());
```

Output
```console
Put Object
request['x-amz-sdk-checksum-algorithm']: 'CRC32'
request['x-amz-checksum-crc32']: 'ShexVg=='
response['x-amz-checksum-crc32']: 'ShexVg=='
response['x-amz-checksum-type']: 'FULL_OBJECT'

Get Object
request['x-amz-checksum-mode']: 'ENABLED'
response['x-amz-checksum-crc32']: 'ShexVg=='
response['x-amz-checksum-type']: 'FULL_OBJECT'
```

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/middleware-flexible-checksums/src/flexibleChecksumsInputMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    next: SerializeHandler<any, Output>,
+    context: HandlerExecutionContext
+  ): SerializeHandler<any, Output> =>
+  async (args: SerializeHandlerArguments<any>): Promise<SerializeHandlerOutput<Output>> => {
+    const input = args.input;
+    const { requestValidationModeMember } = middlewareConfig;
+    const responseChecksumValidation = await config.responseChecksumValidation();
+
+    if (requestValidationModeMember && responseChecksumValidation === ResponseChecksumValidation.WHEN_SUPPORTED) {
+      input[requestValidationModeMember] = "ENABLED";
```
Comment (kuhe) [inline packages/middleware-flexible-checksums/src/flexibleChecksumsInputMiddleware.ts:53]: what happens if the `if` conditions are true but `input[requestValidationModeMember]` is set by the user to some other value than `ENABLED` (ignoring the fact that this is the only enum value)? 

I would expect us to allow the value through and allow the user to receive a validation exception from the server.

**Review by kuhe — APPROVED** [review]

File: packages/middleware-flexible-checksums/src/flexibleChecksumsInputMiddleware.ts (hunk 1 above)
Comment (trivikr) [reply packages/middleware-flexible-checksums/src/flexibleChecksumsInputMiddleware.ts:53]: The `input[requestValidationModeMember]` is only allowed to be `ENABLED` as per specification.

This is what the requirements say
> Check if the user set requestValidationModeMember to ENABLED:
> * If they did, continue.
> * If they didn’t, check the value of ResponseChecksumValidation:
>   * If it's set to WHEN_SUPPORTED, set requestValidationModeMember to ENABLED and continue.
>   * If it's set to WHEN_REQUIRED, go to END.

To strictly satisfy the requirements, we should to check `input[requestValidationModeMember]` is not set to `ENABLED`. But it doesn't matter, since the value will be overridden to be `ENABLED` anyway.

**Comment by trivikr** [comment]
The default checksum is changed from CRC32 to SHA256 based on benchmarks provided in https://github.com/aws/aws-sdk-js-v3/issues/6499

**Review by kuhe — APPROVED** [review]

**Comment by trivikr** [comment]
The default checksum is switched back to CRC32 based on internal discussions

**Review by kuhe — APPROVED** [review]

**Review by kuhe — APPROVED** [review]

**Comment by trivikr** [comment]
The checksum computation is not skipped for `x-amz-checksum-algorithm` in [`9204de2` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/9204de2eee5bbf914b2388db4457ced7352359d3)

**Comment by trivikr** [comment]
Commit to set input[requestAlgorithmMember] to default checksum algorithm if it's not set and either requestChecksumCalculation is supported or requestChecksumRequired is set to true: [`317926c` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/317926c14ba8051ae6b0bcfbf8f98ac80c3021bb)

**Comment by trivikr** [comment]
Commit to remove the case from flexibleChecksumsMiddleware where the input[requestAlgorithmMember] is not set, as that case is now handled in flexibleChecksumsInputMiddleware: [`dc5b5ae` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/dc5b5aec481b2ea8af41e8e8b9b906be5900534c)

File: packages/middleware-flexible-checksums/src/constants.ts [hunk 2]
```diff
@@ -74,8 +74,3 @@ export enum ChecksumLocation {
  * @internal
  */
 export const DEFAULT_CHECKSUM_ALGORITHM = ChecksumAlgorithm.CRC32;
-
-/**
- * @internal
- */
-export const S3_EXPRESS_DEFAULT_CHECKSUM_ALGORITHM = ChecksumAlgorithm.CRC32;
```
Comment (kuhe) [inline packages/middleware-flexible-checksums/src/constants.ts:78]: is it possible S3 will change the "default algorithm" again while intending S3 Express to stay on CRC32?

**Review by kuhe — APPROVED** [review]

**Comment by trivikr** [comment]
The tests are failing because the input is processed from `middlewareConfig` sent from and not `args`.
Since it's not related to using RequestChecksumCalculation and ResponseChecksumValidation, I'll fix it outside of PR and rebase.

**Comment by trivikr** [comment]
100+ integration tests were added in [`ae9d5e9` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/ae9d5e907c3c7f256fafe405109ad64c1c1d4e2b)

We also fixed an edge case to [skip checksum if input[requestAlgorithmMember] is not set](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/2b17b51a5316f852b16352ed5b23732ba6c0d3fc)

**Comment by trivikr** [comment]
An edge case to leave input.requestAlgorithmMember if checksum header is set was implemented in [`969be81` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/969be814e10cf17d8c95d66ee769e9f0d4c3e0ca)

**Comment by trivikr** [comment]
To handle previous ask, the input[requestAlgorithmMember] (if added by SDK) was removed in [`5f86077` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/5f86077d82fcbf26666138d076d168c6378c5a09) if checksum header is provided by the user.

File: packages/middleware-flexible-checksums/src/flexibleChecksumsMiddleware.ts [hunk 3]
```diff
@@ -54,15 +54,25 @@ export const flexibleChecksumsMiddleware =
       return next(args);
     }
 
-    if (hasHeaderWithPrefix("x-amz-checksum-", args.request.headers)) {
+    const { request, input } = args;
+    const { body: requestBody, headers } = request;
+    const { requestChecksumRequired, requestAlgorithmMember } = middlewareConfig;
+
+    if (hasHeaderWithPrefix("x-amz-checksum-", headers)) {
+      // Remove input[requestAlgorithmMember] and header, if it was added by flexibleChecksumsInputMiddleware
```
Comment (kuhe) [inline packages/middleware-flexible-checksums/src/flexibleChecksumsMiddleware.ts:62]: this should state the condition more completely, since 

> Remove input[requestAlgorithmMember] and header, if it was added by flexibleChecksumsInputMiddleware

sounds like an unconditional removal 

it should perhaps say "... and the user has set the checksum via a modeled field"?

**Comment by trivikr** [comment]
Instead of populating input[requestAlgorithmMember] in InputMiddleware and attempting to remove it from main middleware, we switched to using newly introduced `requestAlgorithmMemberHttpHeader` to set httpHeader in [`b614dbe` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/b614dbed2151b2b896e67792029ce0582d3bfb05)

**Comment by trivikr** [comment]
Instead of populating input[requestAlgorithmMember] in InputMiddleware and attempting to remove it from main middleware, we switched to using newly introduced `requestAlgorithmMemberHttpHeader` to set httpHeader in [`b614dbe` (#6492)](https://github.com/aws/aws-sdk-js-v3/pull/6492/commits/b614dbed2151b2b896e67792029ce0582d3bfb05)

**Comment by trivikr** [comment]
This PR is open since Sep'24 and merge is delayed further.
The rebase has been difficult since there were many tooling changes in JS SDK v3.

I've created two new PRs based on the tasks they perform and with squashed commits
* https://github.com/aws/aws-sdk-js-v3/pull/6749
* https://github.com/aws/aws-sdk-js-v3/pull/6750
